R/coerce.R

```diff
@@ -0,0 +1,180 @@
+#'   can be provided as a numeric value, numeric vector, list, or file path to a
+#'   JSON file for the 10x Genomics Visium platform. For 10x Genomics Visium,
+#'   the correct scale factor will automatically be selected depending on the
+#'   resolution of the image from \code{imageSources}. Default = \code{1}.
+#' @param imgData Optional \code{\link{DataFrame}} containing the image data.
+#'   Alternatively, this can be built from the arguments \code{imageSources} and
+#'   \code{image_id} (see Details).
+#' @param imageSources Optional file path(s) or URL(s) for one or more image
+#'   sources.
+#' @param image_id Optional character vector (same length as
+#'   \code{imageSources}) containing unique image identifiers.
+#' @param loadImage Logical indicating whether to load image into memory.
+#'   Default = \code{FALSE}.
+
+#' @aliases 
+#' coerce, SingleCellExperiment, SpatialExperiment-method
+#' toSpatialExperiment
+#' 
+#' @examples 
+#' dir <- system.file(
+#' file.path("extdata", "10xVisium", "section1"),
+#' package = "SpatialExperiment")
+#' 
+#' # read in counts
+#' fnm <- file.path(dir, "raw_feature_bc_matrix")
+#' sce <- DropletUtils::read10xCounts(fnm)
+#' 
+#' 
+#' # read in spatial coordinates
+#' fnm <- file.path(dir, "spatial", "tissue_positions_list.csv")
+#' xyz <- read.csv(fnm, header = FALSE,
+#'                 col.names = c(
+#'                     "barcode", "in_tissue", "array_row", "array_col",
+#'                     "pxl_row_in_fullres", "pxl_col_in_fullres"))
+#' 
+#' (spe <- as(sce, "SpatialExperiment"))
+#' ## as method
+#' int_colData(sce)$spatialData <- DataFrame(xyz[,c(1:4)])
+#' int_colData(sce)$spatialCoords <- as.matrix(xyz[,c(5,6)])
+#' ## Coercing an sce without imgData
+#' (spe <- as(sce, "SpatialExperiment"))
+#' ## Coercing an sce with imgData
+#' img <- readImgData(
+#'     path = file.path(dir, "spatial"),
+#'     sample_id = "sample01")
+#' int_colData(sce)$imgData <- img
+#' (spe <- as(sce, "SpatialExperiment"))
+#' 
+#' ## toSpatialExperiment method
+#' (spe <- toSpatialExperiment(sce,
+#'     imgData = img,
+#'     spatialData = DataFrame(xyz), 
+#'     spatialCoordsNames = c("pxl_col_in_fullres", "pxl_row_in_fullres"),
+#'     sample_id = "sample01"))
+NULL
+
+
+setAs(
+    from="SingleCellExperiment", 
+    to="SpatialExperiment", 
+    function(from) {
+        spatialCoords <- spatialData <- imgData <- NULL
+        sample_id <- "sample01"
+        if ("spatialCoords" %in% names(int_colData(from))) {
+            spatialCoords <- int_colData(from)$spatialCoords
+        }
+        if ("spatialData" %in% names(int_colData(from))) {
+            spatialData <- int_colData(from)$spatialData
+        }
+        if ("imgData" %in% names(int_colData(from))) {
+            imgData <- int_colData(from)$imgData
+        }
+        if ("sample_id" %in% colnames(colData(from))) {
+            sample_id <- unique(colData(from)$sample_id)
+        }
```

Comment:
lines 116-129 could be greatly simplified by
```
spatialCoords <- int_colData(from)$spatialCoords
spatialData <- int_colData(from)$spatialData
imgData <- int_colData(from)$imgData
sample_id <- unique(from$sample_id)
```
All of these will give NULL if the slot doesn't exist, and the slot's value if it does.

Reply:
Maybe I wasn't clear... but this could be
```
setAs(
    from="SingleCellExperiment", 
    to="SpatialExperiment", 
    function(from) {
sample_id <- unique(from$sample_id)
if (is.null(sample_id)) sample_id <- "sample01"
icd <- int_colData(from)
spe <- .sce_to_spe(from, 
  sample_id=sample_id,
  spatialData=icd$spatialData,
  spatialCoords=icd$spatialCoords,
  imgData=icd$imgData)
return(spe)
    }
)
```

Reply:
ok, done! :)